Dev table persistence

Briggs-Richardson opened the pull request:
DO NOT MERGE.

Briggs-Richardson committed:
Creates encoding table data persistence using QSettings. If a session is saved, then the user can decide if the previous table data is loaded at the start of the application.

Briggs-Richardson committed:
Adds a "Welcome" dialog box for session selection

This patch adds functionality to save user data across multiple sessions. This
functionality is achieved using the QSettings QT class. Sessions are stored in
groups, with group names acting as session identifiers. Sessions are
automatically saved or updated when the application closes. The identifier of a
session if chosen to be the time recorded when a new session is exited for the
first time. The naming of a session ID may change later on.

With this functionality, a "Welcome" dialog box is introduced, displaying a
selection of saved sessions that the user can choose from, including
"New Session". If the user selects a New Session, it will be remembered
for future use. If the user selects a saved session, then any changes will
update the selected session.

Testing Steps
  1. Run main.py to start the application.
  2. Verify that the "Welcome" dialog box opens, and contains a "New
     Session", "Ok", and "Cancel" buttons.
  3. Verify that clicking the "Cancel" button closes the dialog box and
     that the main application does not open.
  4. Verify that clicking the "Ok" button without a Session button
     selected does not produce any effects.
  5. Click on the "New Session" option and click the "Ok" button.
  6. Verify that the main application window appears.
  7. Verify that there is no existing table data.
  8. Record "123" into the first table cell in the first row.
  9. Close the application and open it again.
  10. Verify that the "Welcome" dialog box contains a button for a "New
      Session" and a new session with text that represents the time that
      the application was recently closed.
  11. Select the saved session and click "Ok".
  12. Verify that the "123" text in the first cell of the first row
      exists.
  13. Record "456" into the second cell of the first row and close the
      application.
  14. Run the application, and verify that the dialog box presents no
      new session identifier options.
  15. Load the saved session again and verify that the "123" and "456"
      cells exist.
  16. Repeat steps 5-15 to test a second saved session, this time
      entering in different cell values of your choice.
  17. Open the application and select the first saved session again, and
      verify that the "123" and "456" cells exist.

Type: New Feature

Briggs-Richardson committed:
Merge branch 'main' into dev_table_persistence

Jcobo314 commented on View/main_window.py:
```diff
@@ -106,5 +98,11 @@ def write_settings(self):
         Saves the state of the application to the QSettings object.
         """
         settings = QSettings()
```
Settings not used so line can be removed

Briggs-Richardson committed:
Adds a "Welcome" dialog box for session selection

This patch adds functionality to save user data across multiple sessions. This
functionality is achieved using the QSettings QT class. Sessions are stored in
groups, with group names acting as session identifiers. Sessions are
automatically saved or updated when the application closes. The identifier of a
session if chosen to be the time recorded when a new session is exited for the
first time. The naming of a session ID may change later on.

With this functionality, a "Welcome" dialog box is introduced, displaying a
selection of saved sessions that the user can choose from, including
"New Session". If the user selects a New Session, it will be remembered
for future use. If the user selects a saved session, then any changes will
update the selected session.

Testing Steps
  1. Run main.py to start the application.
  2. Verify that the "Welcome" dialog box opens, and contains a "New
     Session", "Ok", and "Cancel" buttons.
  3. Verify that clicking the "Cancel" button closes the dialog box and
     that the main application does not open.
  4. Verify that clicking the "Ok" button without a Session button
     selected does not produce any effects.
  5. Click on the "New Session" option and click the "Ok" button.
  6. Verify that the main application window appears.
  7. Verify that there is no existing table data.
  8. Record "123" into the first table cell in the first row.
  9. Close the application and open it again.
  10. Verify that the "Welcome" dialog box contains a button for a "New
      Session" and a new session with text that represents the time that
      the application was recently closed.
  11. Select the saved session and click "Ok".
  12. Verify that the "123" text in the first cell of the first row
      exists.
  13. Record "456" into the second cell of the first row and close the
      application.
  14. Run the application, and verify that the dialog box presents no
      new session identifier options.
  15. Load the saved session again and verify that the "123" and "456"
      cells exist.
  16. Repeat steps 5-15 to test a second saved session, this time
      entering in different cell values of your choice.
  17. Open the application and select the first saved session again, and
      verify that the "123" and "456" cells exist.

Type: New Feature
